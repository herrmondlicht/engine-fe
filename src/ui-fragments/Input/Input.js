import React from "react";
import { Label } from "ui-fragments/Typography";
import { css } from "@emotion/css";

const inputBaseClass = [
  "p-4",
  "bg-gray-100",
  "placeholder-gray-400",
  "rounded-2xl",
  "text-gray-800",
  "focus:bg-white",
  "focus:ring-2",
  "focus:ring-primary-0",
].join(" ");

const inputError = ["bg-error-2", "ring-2", "ring-error-0"].join(" ");

const placeholderClass = css`
  ::-webkit-input-placeholder {
    /* WebKit browsers */
    text-transform: none;
  }
  :-moz-placeholder {
    /* Mozilla Firefox 4 to 18 */
    text-transform: none;
  }
  ::-moz-placeholder {
    /* Mozilla Firefox 19+ */
    text-transform: none;
  }
  :-ms-input-placeholder {
    /* Internet Explorer 10+ */
    text-transform: none;
  }
  ::placeholder {
    /* Recent browsers */
    text-transform: none;
  }
`;

const Input = React.forwardRef(
  ({ label, fw, placeholder, error, uppercase, ...props }, ref) => {
    const modifierClasses = [
      ...(fw ? ["w-full"] : []),
      ...(error ? [inputError] : []),
      ...(uppercase ? ["uppercase"] : []),
    ].join(" ");
    return (
      <div className="pb-6 relative">
        {label ? (
          <div className="mb-4 ml-1">
            <Label error={error}>{label}</Label>
          </div>
        ) : null}
        <input
          autoComplete="no"
          placeholder={placeholder}
          ref={ref}
          {...props}
          className={`${inputBaseClass} ${modifierClasses} ${placeholderClass}`}
        />
        {error && (
          <label className="absolute w-full bottom-0 left-0 font-medium ml-1 mt-1 text-sm text-error-0">
            {error}
          </label>
        )}
      </div>
    );
  }
);

export { Input };
