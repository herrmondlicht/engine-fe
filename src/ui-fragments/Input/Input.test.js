import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import { Input } from "./Input";
import userEvent from "@testing-library/user-event";

describe("Fragment/Input", () => {
  it("Renders the input", async () => {
    render(<Input />);
    expect(screen.getByRole("textbox")).toMatchSnapshot();
  });

  it("Renders the input as textarea", async () => {
    render(<Input as="textarea" />);
    expect(screen.getByRole("textbox")).toMatchSnapshot();
  });

  it("Renders the input with initial value", async () => {
    const INITIAL_VALUE = "Hey there!";
    render(<Input value={INITIAL_VALUE} onChange={e => e} />);
    expect(screen.getByRole("textbox")).toHaveValue(INITIAL_VALUE);
  });

  it("Changes the value of the input", async () => {
    const TYPE_MESSAGE = "How are you doing?";
    render(<Input />);
    const renderedInput = screen.getByRole("textbox");

    await userEvent.type(renderedInput, TYPE_MESSAGE);

    expect(renderedInput).toHaveValue(TYPE_MESSAGE);
  });

  it("Renders the label", async () => {
    const LABEL_TEXT = "How are you doing?";
    render(<Input label={LABEL_TEXT} />);
    expect(screen.getByText(LABEL_TEXT)).toBeDefined();
  });

  it("Show error indicators", async () => {
    const LABEL_TEXT = "How are you doing?";
    const ERROR_TEXT = "The computer needs you to be fine";
    render(<Input label={LABEL_TEXT} error={ERROR_TEXT} />);
    expect(screen.getByRole("textbox")).toMatchSnapshot();
    expect(screen.getByText(ERROR_TEXT)).toBeDefined();
  });
});
