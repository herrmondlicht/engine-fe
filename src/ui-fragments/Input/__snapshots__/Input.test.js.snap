// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fragment/Input Renders the input 1`] = `
<input
  autocomplete="no"
  class="p-4 bg-gray-100 placeholder-gray-400 rounded-2xl text-gray-800 focus:bg-white focus:ring-2 focus:ring-primary-0 resize-none  css-1bya8sb css-9bnagd"
/>
`;

exports[`Fragment/Input Renders the input as textarea 1`] = `
<textarea
  autocomplete="no"
  class="p-4 bg-gray-100 placeholder-gray-400 rounded-2xl text-gray-800 focus:bg-white focus:ring-2 focus:ring-primary-0 resize-none  css-1bya8sb css-3cn25m"
/>
`;

exports[`Fragment/Input Show error indicators 1`] = `
<input
  autocomplete="no"
  class="p-4 bg-gray-100 placeholder-gray-400 rounded-2xl text-gray-800 focus:bg-white focus:ring-2 focus:ring-primary-0 resize-none bg-error-2 ring-2 ring-error-0 css-1bya8sb css-9bnagd"
/>
`;
